File: javascript/1394.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var findLucky = function(arr) {
    let obj = {};
    let res = -1;
    for(let num of arr) {
        if(obj[num]) obj[num]++;
        else obj[num] = 1;
    }
    for(let key in obj) {
        if(obj[key] == key) res = (res < key) ? obj[key] : res;
    }
    return res;
};

console.log(findLucky([2,2,3,4]));
console.log(findLucky([1,2,2,3,3,3]));
console.log(findLucky([2,2,2,3,3]));

File: cpp/1394.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int findLucky(vector<int>& arr) {
        unordered_map<int, int> freq;
        int res = -1;
        for (int num : arr)
            freq[num]++;
        for (auto& [n, f] : freq) {
            if (n == f) res = max(res, n);
        }
        return res;
    }
};
